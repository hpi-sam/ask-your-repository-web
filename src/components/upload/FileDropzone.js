// @flow
import React from 'react';
import Dropzone from 'react-dropzone';
import type { DropFilesEventHandler } from 'react-dropzone';
import './FileDropzone.scss';

type Props = {
  onDrop: DropFilesEventHandler,
};

const FileDropzone = ({ onDrop }: Props) => (
  <Dropzone
    className="FileDropzone"
    activeClassName="FileDropzone--active"
    acceptClassName="FileDropzone--accept"
    disabledClassName="FileDropzone--disabled"
    onDrop={onDrop}
    data-cy="upload-dropzone"
  >
    <span className="FileDropzone__text">
      Place your file(s) here
    </span>
  </Dropzone>
);

export default FileDropzone;
