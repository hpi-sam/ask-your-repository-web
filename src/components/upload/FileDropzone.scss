@import '../../style/breakpoints.scss';

.FileDropzone {
  --bg-color: transparent;
  --hover-bg-color: rgba(255, 255, 255, .03);

  --border-width: 4px;
  --border-color: #fff;
  --hover-border-color: rgb(36, 186, 255);

  --text-color: #fff;
  --hover-text-color: rgb(36, 186, 255);

  --transition-duration: 0.2s;

  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

  height: 100%;
  width: 100%;
  padding: 20px;

  background-color: var(--bg-color);
  border: var(--border-width) dashed var(--border-color);
  border-radius: inherit;

  transition:
    background-color var(--transition-duration),
    border-color var(--transition-duration),
    color var(--transition-duration);

  &:hover {
    --bg-color: var(--hover-bg-color);
    --text-color: var(--hover-text-color);
    --border-color: var(--hover-border-color);
  }

  &__text {
    text-align: center;
    font-weight: bold;
    letter-spacing: 3px;
    text-transform: uppercase;
    color: var(--text-color);
  }
}
