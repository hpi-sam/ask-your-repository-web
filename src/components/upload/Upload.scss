@import '../../style/breakpoints.scss';

.Upload {
  display: flex;
  width: 100%;
  min-height: 100%;
  max-height: 100%;
  margin: 20px 0;
  padding: 0 30px;

  &__tagging {
    flex: 1;
    margin-right: 2em;
    overflow: hidden;
  }

  &__uploads {
    display: flex;
    flex-direction: column;
    max-height: 100%;
    margin-left: auto;
    margin-right: 0;
  }

  &__save-disabled-hint {
    font-size: .8em;
    color: #666;
    font-weight: bold;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 1em;
  }

  &__save-button {
    align-items: center;
    text-transform: uppercase;
    font-weight: bold;
    margin-left: auto;
    margin-right: auto;
    padding: 1rem 2rem;
    border-radius: .3rem;

    &__text {
      display: none;

      @media (min-width: $xs-to-sm) {
        display: inline-flex;
      }
    }

    &__icon {
      @media (min-width: $xs-to-sm) {
        margin-right: 1rem;
      }
    }
  }
}
