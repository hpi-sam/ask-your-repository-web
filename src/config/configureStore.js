// @flow
import { applyMiddleware, createStore } from 'redux';
import type { Dispatch } from 'redux';
import { persistStore, persistReducer } from 'redux-persist-erksch';
import storage from 'redux-persist-erksch/lib/storage';
import { routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { middleware as flashMiddleware } from 'redux-flash';
import thunk from 'redux-thunk';
import socketioMiddleware from '../middleware/socketioMiddleware';
import axiosAuthMiddleware from './axiosAuthMiddleware';
import createRootReducer from '../state/rootReducer';
import type { AppState } from '../state/AppState';
import type { Action } from '../state/Action';

export const history = createBrowserHistory();

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['activeTeam', 'presentationMode', 'auth'],
};

function configureStore() {
  let middleware = [
    thunk,
    flashMiddleware(),
    routerMiddleware(history),
    socketioMiddleware(),
    axiosAuthMiddleware(),
  ];

  if (process.env.NODE_ENV === 'development') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  const rootReducer = createRootReducer(history);
  const persistedReducer: any = persistReducer(persistConfig, rootReducer);

  const store = createStore<AppState, Action, Dispatch<Action>>(
    persistedReducer,
    applyMiddleware(...middleware),
  );

  const persistor = persistStore(store);

  return { store, persistor };
}

export default configureStore;
