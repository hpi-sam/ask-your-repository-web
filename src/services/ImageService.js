// @flow
import humps from 'humps';
import api from '../config/api';
import type { Image } from '../models/Image';
import type { Tag } from '../models/Tag';

type ListParams = {
  teamId: string,
  offset?: number,
  limit?: number,
  search?: string,
};

class ImageService {
  static async get(id: string): Promise<Image> {
    const response = await api.get(`/images/${id}`);
    const image: any = humps.camelizeKeys(response.data);
    return image;
  }

  static async list(params: ListParams): Promise<Image[]> {
    const response = await api.get('/images', {
      params: humps.decamelizeKeys(params),
    });
    const images: any = humps.camelizeKeys(response.data.images);
    return images;
  }

  static async related(id: string, limit: number = 5) {
    const response = await api.get(`/images/${id}/related`, {
      params: { limit },
    });
    const images: any = humps.camelizeKeys(response.data.images);
    return images;
  }

  static async create(formData: FormData): Promise<Image> {
    const headers = { 'Content-Type': 'multipart/form-data' };
    const response = await api.post('/images', formData, { headers });
    const image: any = humps.camelizeKeys(response.data);
    return image;
  }

  static async addTags(imageId: string, tags: Array<Tag>): Promise<Image> {
    const response = await api.post(`/images/${imageId}/tags`, { tags });
    const image: any = humps.camelizeKeys(response.data);
    return image;
  }

  static async patch(imageId: string, data: Object) {
    await api.patch(`/images/${imageId}`, data);
  }

  static async delete(imageId: string) {
    await api.delete(`images/${imageId}`);
  }

  static async patchMany(images: Array<{ id: string, tags: Array<Tag> }>) {
    await api.patch('/images', { artifacts: images });
  }
}

export default ImageService;
