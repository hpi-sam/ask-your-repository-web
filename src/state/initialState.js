// @flow

// This file is only for state related testing

import type { AppState } from './AppState';
import { initialState as presentation } from './presentation/presentation.reducer';
import { initialState as teamSidebar } from './team_sidebar/teamSidebar.reducer';
import { initialState as presentationMode } from './presentation_mode/presentationMode.reducer';
import { initialState as activeTeam } from './active_team/activeTeam.reducer';
import { initialState as auth } from './auth/auth.reducer';

const initialState: AppState = {
  flash: { messages: [] },
  presentation,
  activeTeam,
  teamSidebar,
  presentationMode,
  auth,
};

export default initialState;
