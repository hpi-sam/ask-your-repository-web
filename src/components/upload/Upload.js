// @flow
import React, { Component, Fragment } from 'react';
import uuidv4 from 'uuid/v4';
import { push } from 'connected-react-router';
import { connect } from 'react-redux';
import { FiSave } from 'react-icons/fi';
import FileDropzone from './FileDropzone';
import ImageService from '../../services/ImageService';
import SaveButton from '../utility/SaveButton';
import UploadList from './UploadList';
import UploadKeyboardListener from './UploadKeyboardHandler';
import type { Image } from '../../models/Image';
import type { Team } from '../../models/Team';
import type { Upload as UploadType } from '../../models/Upload';
import type { AppState } from '../../state/AppState';
import './Upload.scss';
import MultiTagging from '../tagging/MultiTagging';

type Props = {
  dispatch: Function,
  activeTeam: ?Team,
};

type State = {
  uploads: Array<UploadType>,
  selectedUploadId: ?string,
};

class Upload extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      uploads: [],
      selectedUploadId: null,
    };
  }

  getImages(): Array<?Image> {
    return this.state.uploads.map(upload => upload.image);
  }

  getSuccessfulImages() {
    return this.getImages().filter(image => !!image);
  }

  getUploadOfImage(imageId: string): ?UploadType {
    return this.state.uploads
      .find(upload => upload.image && upload.image.id === imageId);
  }

  addUploads = (uploads: UploadType[]) => {
    this.setState(state => ({
      uploads: [...state.uploads, ...uploads],
    }));
  };

  updateUpload = (id: string, updateData: Object) => {
    const { uploads } = this.state;
    const index = uploads.findIndex(upload => upload.id === id);
    uploads[index] = { ...uploads[index], ...updateData };

    this.setState({ uploads });
  };

  getSelectedUpload = (): ?UploadType => {
    const { uploads, selectedUploadId } = this.state;
    return uploads.find(upload => upload.id === selectedUploadId);
  };

  setSelectedUpload = (selectedUploadId: string) => {
    this.setState({ selectedUploadId });
  };

  hasSelectedUpload = () => !!this.state.selectedUploadId;

  handleFileDrop = (files: File[]) => {
    if (files.length === 0) return;

    const newUploads = files.map(file => ({
      file,
      status: 'ongoing',
      id: uuidv4(),
      image: null,
    }));

    if (!this.hasSelectedUpload()) {
      this.setSelectedUpload(newUploads[0].id);
    }

    this.addUploads(newUploads);

    newUploads.forEach((upload) => {
      this.submitImage(upload.file)
        .then(image => this.updateUpload(upload.id, {
          status: 'succeeded',
          image,
        }))
        .catch(() => this.updateUpload(upload.id, {
          status: 'failed',
        }));
    });
  };

  handleSubmitClick = async () => {
    const patchData = this.getSuccessfulImages()
      .map(image => ({
        id: image.id,
        tags: image.tags,
      }));

    await ImageService.patchMany(patchData);
    this.props.dispatch(push('/images'));
  };

  handleImageTagsChange = (imageId: string, tags: Array<Tag>) => {
    const upload = this.getUploadOfImage(imageId);

    this.updateUpload(upload.id, {
      image: { ...upload.image, tags },
    });
  };

  submitImage(file: File): Promise<Image> {
    const { activeTeam } = this.props;

    if (!activeTeam) throw Error('Missing active team.');

    const formData = new FormData();
    formData.append('image', file);
    formData.append('team_id', activeTeam.id);

    return ImageService.create(formData);
  }

  isFormSaveable() {
    const images = this.getSuccessfulImages();
    return images.every(image => image.tags.length !== 0);
  }

  render() {
    const { uploads, selectedUploadId } = this.state;
    const selectedUpload = this.getSelectedUpload();

    const isFormSaveable = this.isFormSaveable();

    return (
      <div className="Upload">
        <UploadKeyboardListener
          uploads={uploads}
          selectedUploadId={selectedUploadId}
          onSelectedChange={this.setSelectedUpload}
        />
        {uploads.length > 0 ? (
          <Fragment>
            <div className="Upload__tagging">
              {selectedUpload && selectedUpload.image && (
                <MultiTagging
                  images={this.getSuccessfulImages()}
                  selectedImageId={selectedUpload.image.id}
                  onImageTagsChange={this.handleImageTagsChange}
                />
              )}
            </div>
            <div className="Upload__uploads">
              <UploadList
                onDrop={this.handleFileDrop}
                uploads={uploads}
                onItemClick={this.setSelectedUpload}
                selectedUploadId={selectedUploadId}
              />
              {!isFormSaveable && (
                <div className="Upload__save-disabled-hint">
                  Add at least one tag to your image(s).
                </div>
              )}
              <SaveButton
                onClick={this.handleSubmitClick}
                className="Upload__save-button"
                disabled={!isFormSaveable}
              >
                <FiSave className="Upload__save-button__icon" />
                <span className="Upload__save-button__text">
                  Finish
                </span>
              </SaveButton>
            </div>
          </Fragment>
        ) : (
          <FileDropzone onDrop={this.handleFileDrop} />
        )}
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  activeTeam: state.activeTeam,
});

export default connect(mapStateToProps)(Upload);
