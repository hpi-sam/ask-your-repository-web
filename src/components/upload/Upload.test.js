// @flow
import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Upload from './Upload';
import FileDropzone from './FileDropzone';
import initialState from '../../state/initialState';
import TeamFactory from '../../factories/TeamFactory';
import ImageService from '../../services/ImageService';

const mockStore = configureStore();

jest.mock('../../services/ImageService', () => {
  const image = { id: 1, tags: [] };

  return {
    create: jest.fn(() => Promise.resolve(image)),
  };
});

jest.mock('../../services/TagService', () => ({
  suggested: jest.fn(() => Promise.resolve([])),
}));

describe('<Upload />', () => {
  let wrapper;
  let wrapperInstance;
  let store;

  const state = { ...initialState, activeTeam: TeamFactory.createStaticDummyTeam() };

  beforeEach(() => {
    store = mockStore(state);
    wrapper = mount((
      <Provider store={store}>
        <Upload />
      </Provider>
    )).find('Upload');
    wrapperInstance = wrapper.instance();
  });

  afterEach(() => {
    ImageService.create.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders an FileDropzone component', () => {
    expect(wrapper.find(FileDropzone).exists()).toBeTruthy();
  });

  describe('an image has been dropped', () => {
    const imageFile = new File([], 'goat.jpg', { type: 'image/jpeg' });

    beforeEach(async () => {
      await wrapperInstance.handleFileDrop([imageFile]);
    });

    it('should call image create', () => {
      expect(ImageService.create).toHaveBeenCalled();
    });

    it('should send the dropped file as form-data', async () => {
      const formData = ImageService.create.mock.calls[0][0];
      const sentFile = formData.get('image');

      let sentFileData: string;
      let imageBlobData: string;

      function asyncFileReader(file: File, onLoad: (string) => void) {
        return new Promise((resolve) => {
          const fileReader = new FileReader();
          fileReader.onload = (e) => {
            onLoad(e.target.result);
            resolve();
          };
          fileReader.readAsDataURL(file);
        });
      }

      await Promise.all([
        asyncFileReader(sentFile, (result) => { sentFileData = result; }),
        asyncFileReader(imageFile, (result) => { imageBlobData = result; }),
      ]);

      expect(sentFileData).toEqual(imageBlobData);
    });

    describe('upload failed because of network error', () => {
      beforeAll(() => {
        ImageService.create.mockImplementation(() => Promise.reject(Response.error()));
      });

      it('should catch the error', () => {});
    });

    describe('upload failed because of server error', () => {
      const errorMessage = 'Dummy Error';
      const errorResponse = { message: errorMessage };

      beforeAll(() => {
        ImageService.create.mockImplementation(() => Promise.reject({ response: errorResponse }));
      });

      it('should catch the error', () => {});
    });
  });

  it('do nothing if image drop is called without any files', () => {
    wrapperInstance.handleFileDrop([]);
    expect(ImageService.create).not.toHaveBeenCalled();
  });
});
