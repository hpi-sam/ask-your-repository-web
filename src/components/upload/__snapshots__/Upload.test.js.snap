// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Upload /> renders correctly 1`] = `
<Upload
  activeTeam={
    Object {
      "id": "2dedf601-f0bf-4290-ac3c-f5761234d068",
      "name": "odio expedita aut",
    }
  }
  dispatch={[Function]}
>
  <div
    className="Upload"
  >
    <UploadKeyboardListener
      onSelectedChange={[Function]}
      selectedUploadId={null}
      uploads={Array []}
    />
    <FileDropzone
      onDrop={[Function]}
    >
      <n
        acceptClassName="FileDropzone--accept"
        activeClassName="FileDropzone--active"
        className="FileDropzone"
        data-cy="upload-dropzone"
        disableClick={false}
        disabled={false}
        disabledClassName="FileDropzone--disabled"
        getDataTransferItems={[Function]}
        inputProps={Object {}}
        maxSize={Infinity}
        minSize={0}
        multiple={true}
        onDrop={[Function]}
        preventDropOnDocument={true}
      >
        <div
          aria-disabled={false}
          className="FileDropzone"
          data-cy="upload-dropzone"
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <span
            className="FileDropzone__text"
          >
            Place your file(s) here
          </span>
          <input
            autoComplete="off"
            disabled={false}
            multiple={true}
            onChange={[Function]}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 0.00001,
                "pointerEvents": "none",
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            type="file"
          />
        </div>
      </n>
    </FileDropzone>
  </div>
</Upload>
`;
